sdk/data/azcosmos/cosmos_global_endpoint_manager_test.go

```diff
@@ -0,0 +1,407 @@
+		return
+	}
+	req.SetOperationValue(pipelineRequestOptions{
+		isWriteOperation: true,
+	})
+
+	client := &Client{endpoint: srv.URL(), pipeline: pl}
+
+	preferredRegions := []string{"West US", "Central US"}
+
+	gem, err := newGlobalEndpointManager(client, preferredRegions, 5*time.Minute)
+	assert.NoError(t, err)
+
+	accountProps, err := gem.GetAccountProperties()
+	assert.NoError(t, err)
+
+	expectedAccountProps := accountProperties{
+		ReadRegions:                  nil,
+		WriteRegions:                 nil,
+		EnableMultipleWriteLocations: false,
+	}
+	assert.Equal(t, expectedAccountProps, accountProps)
+}
+
+func TestGlobalEndpointManagerCanUseMultipleWriteLocations(t *testing.T) {
+	headerPolicy := &headerPolicies{}
+	srv, close := mock.NewTLSServer()
+	defer close()
+	srv.SetResponse(mock.WithStatusCode(http.StatusOK))
+
+	verifier := headerPoliciesVerify{}
+	pl := azruntime.NewPipeline("azcosmostest", "v1.0.0", azruntime.PipelineOptions{PerCall: []policy.Policy{headerPolicy, &verifier}}, &policy.ClientOptions{Transport: srv})
+	req, err := azruntime.NewRequest(context.Background(), http.MethodGet, srv.URL())
+	assert.NoError(t, err)
+
+	req.SetOperationValue(pipelineRequestOptions{
+		isWriteOperation: true,
+	})
+
+	client := &Client{endpoint: srv.URL(), pipeline: pl}
+
+	preferredRegions := []string{"West US", "Central US"}
+
+	serverEndpoint, err := url.Parse(srv.URL())
+	assert.NoError(t, err)
+
+	mockLc := newLocationCache(preferredRegions, *serverEndpoint)
+	mockLc.enableMultipleWriteLocations = true
+	mockLc.useMultipleWriteLocations = true
+
+	mockGem := globalEndpointManager{
+		client:              client,
+		preferredLocations:  preferredRegions,
+		locationCache:       mockLc,
+		refreshTimeInterval: 5 * time.Minute,
+	}
+
+	gem, err := newGlobalEndpointManager(client, preferredRegions, 5*time.Minute)
+	assert.NoError(t, err)
+
+	// Multiple locations should be false for default GEM
+	canUseMultipleWriteLocs := gem.CanUseMultipleWriteLocations()
+	assert.False(t, canUseMultipleWriteLocs)
+
+	// Mock GEM with multiple write locations available should show true
+	canUseMultipleWriteLocs = mockGem.CanUseMultipleWriteLocations()
+	assert.True(t, canUseMultipleWriteLocs)
+}
+
+func TestGlobalEndpointManagerUpdate(t *testing.T) { //This test should be testing for the lock on the update/ refresh mechanism
+	// Create a mock client
+	headerPolicy := &headerPolicies{}
+	srv, close := mock.NewTLSServer()
+	defer close()
+	srv.SetResponse(mock.WithStatusCode(http.StatusOK))
+
+	westRegion := accountRegion{
+		Name:     "West US",
+		Endpoint: srv.URL(),
+	}
+
+	properties := accountProperties{
+		ReadRegions:                  []accountRegion{westRegion},
+		WriteRegions:                 []accountRegion{westRegion},
+		EnableMultipleWriteLocations: false,
+	}
+
+	jsonString, err := json.Marshal(properties)
+	if err != nil {
+		t.Fatal(err)
+	}
+	srv.SetResponse(mock.WithBody(jsonString))
+
+	verifier := headerPoliciesVerify{}
```

Comment:
one way to verify the number of HTTP calls is add a new policy that just counts the number of requests, and if you do multiple concurrent Update calls, the number of HTTP calls should be 1?

Reply:
will keep this unresolved as I work on the update tests